src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,446 @@
+
+  // Delta must be positive.
+  RequireParamValue<double>("delta", [](double x) { return x >= 0.0; }, true,
+      "delta must be non-negative");
+
+  // Delta must be positive.
+  RequireParamValue<int>("epochs", [](int x) { return x > 0; }, true,
+      "epochs must be non-negative");
+
+  // These are the matrices we might use.
+  arma::mat trainingSet;
+  arma::Row<size_t> labels;
+  arma::Row<size_t> rawLabels;
+  arma::mat testSet;
+  arma::Row<size_t> predictedLabels;
+  size_t numClasses;
+
+  // Load data matrix.
+  if (CLI::HasParam("training"))
+    trainingSet = std::move(CLI::GetParam<arma::mat>("training"));
+
+  // Check if the labels are in a separate file.
+  if (CLI::HasParam("training") && CLI::HasParam("labels"))
+  {
+    rawLabels = std::move(CLI::GetParam<arma::Row<size_t>>("labels"));
+    if (trainingSet.n_cols != rawLabels.n_cols)
+    {
+      Log::Fatal << "The labels must have the same number of points as the "
+          << "training dataset." << endl;
+    }
+  }
+  else if (CLI::HasParam("training"))
+  {
+    // Checking the size of training data if no labels are passed.
+    if (trainingSet.n_rows < 2)
+    {
+      Log::Fatal << "Can't get labels from training data since it has less "
+          << "than 2 rows." << endl;
+    }
+
+    // The initial predictors for y, Nx1.
+    rawLabels = arma::conv_to<arma::Row<size_t>>::from(
+        trainingSet.row(trainingSet.n_rows - 1));
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  // Load the model, if necessary.
+  LinearSVMModel* model;
+  if (CLI::HasParam("input_model"))
+  {
+    model = CLI::GetParam<LinearSVMModel*>("input_model");
+  }
+  else
+  {
+    model = new LinearSVMModel();
+  }
+
+  // Now, do the training.
+  if (CLI::HasParam("training"))
+  {
+    data::NormalizeLabels(rawLabels, labels, model->mappings);
+    numClasses = CLI::GetParam<int>("num_classes") == 0 ?
+        model->mappings.n_elem : CLI::GetParam<int>("num_classes");
+    model->svm.Lambda() = lambda;
+    model->svm.Delta() = delta;
+    model->svm.NumClasses() = numClasses;
+
+    if (optimizerType == "lbfgs")
+    {
+      model->svm.FitIntercept() = intercept;
+
+      ens::L_BFGS lbfgsOpt;
+      lbfgsOpt.MaxIterations() = maxIterations;
+      lbfgsOpt.MinGradientNorm() = tolerance;
+
+      Log::Info << "Training model with L-BFGS optimizer." << endl;
+
+      // This will train the model.
+      model->svm.Train(trainingSet, labels, numClasses, lbfgsOpt);
+    }
+    else if (optimizerType == "psgd")
+    {
+      const double stepSize = CLI::GetParam<double>("step_size");
+      const bool shuffle = !CLI::HasParam("shuffle");
+      const size_t maxIt = epochs * trainingSet.n_cols;
+
+      ens::ConstantStep decayPolicy(stepSize);
+
+      #ifdef HAS_OPENMP
+        #ifdef BUILD_TESTS
+        size_t threads = 1;
+        #else
+        size_t threads = omp_get_max_threads();
+        #endif
```

Comment:
I see that the aim here is to use only one thread when we are testing, but this connects the testing code to the binding, and we should aim to keep these as separate as possible.  Why not just use `omp_set_num_threads(1);` in the tests when we need to specify only one thread?  I think that should work.

Reply:
Sorry for the slow work.
I have updated this you can take a look. :)